
# Comprehensive Exploratory Data Analysis for Gokart Rentals

In this notebook, we perform an in-depth exploratory data analysis (EDA) on the gokart rental dataset. The primary goal is to uncover patterns and relationships between various weather conditions and gokart rental counts. By leveraging multiple techniques such as visualizations, correlation matrices, box plots, and scatter plots, we aim to gain valuable insights into the factors influencing gokart rentals. These insights will guide further analysis and model building efforts.

## Introduction to Exploratory Data Analysis

Exploratory Data Analysis (EDA) is a crucial step in any data science project. It involves summarizing the main characteristics of a dataset and visualizing important aspects to discover patterns, spot anomalies, test hypotheses, and check assumptions. EDA helps to understand the data better and lays the groundwork for building robust predictive models.

Each step is detailed with descriptions, code, visualizations, and insights.

## Type of Problem

The problem we are addressing is predicting gokart rental checkouts based on various features such as date, hour, temperature, wind, humidity, etc. This is a supervised learning problem because we have labeled data (i.e., we know the number of rentals for each set of features).

Furthermore, this is a regression problem because the target variable (number of gokart rentals) is continuous rather than categorical. We aim to predict the actual rental count, not classify it into discrete categories.


In [1]:
# Cell 1: Import Libraries
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


# Suppress FutureWarnings
warnings.simplefilter(action='ignore', category=FutureWarning)

# Cell 2: Load and Preprocess Data
We load the data from the CSV file, convert the `Date` column to datetime format, and extract day, month, and year from the date. Additionally, we replace infinite values with NaNs and drop them to ensure the dataset is clean for further analysis.


In [2]:
# Cell 2: Load and Preprocess Data
# Load the data
data = pd.read_csv('train.csv')

# Convert Date to datetime
data['Date'] = pd.to_datetime(data['Date'], format='%d/%m/%Y')

# Extract day, month, and year
data['Day'] = data['Date'].dt.day
data['Month'] = data['Date'].dt.month
data['Year'] = data['Date'].dt.year

# Replace infinite values with NaN and then drop them
data.replace([np.inf, -np.inf], np.nan, inplace=True)
data.dropna(inplace=True)


# Cell 3: Basic Descriptive Statistics
This cell provides basic descriptive statistics for the dataset, including count, mean, standard deviation, min, max, and quartiles for each numerical column. This helps us understand the overall distribution and characteristics of the data.


In [3]:
# Cell 3: Basic Descriptive Statistics
# Display basic descriptive statistics for the dataset
print("Descriptive Statistics:\n", data.describe())


Descriptive Statistics:
                              Date         Hour  Temperature(C)  Humidity(%)  \
count                        5000  5000.000000     5000.000000  5000.000000   
mean   2018-05-30 07:51:10.080000    11.549200       12.807080    58.254600   
min           2017-12-01 00:00:00     0.000000      -17.500000     0.000000   
25%           2018-02-28 00:00:00     6.000000        3.300000    42.000000   
50%           2018-05-29 00:00:00    12.000000       13.700000    57.000000   
75%           2018-08-29 00:00:00    18.000000       22.500000    74.000000   
max           2018-11-30 00:00:00    23.000000       39.400000    98.000000   
std                           NaN     6.930309       11.997744    20.300921   

       Wind speed (m/s)  Visibility (10m)  Dew point temperature(C)  \
count       5000.000000        5000.00000               5000.000000   
mean           1.726000        1432.21500                  3.995940   
min            0.000000          27.00000         

### Insights from Descriptive Statistics and Data Analysis

#### General Observations:
- The dataset contains 5000 observations, each with multiple weather-related features and the target variable, "Gokart rental count."
- The observations are evenly distributed across different hours of the day, days of the month, and months of the year, mostly covering the year 2018.

#### Weather Conditions:
1. **Temperature**:
   - The average temperature is around 12.8°C, with temperatures ranging from -17.5°C to 39.4°C.
   - This wide range indicates the data includes observations from various seasons and climatic conditions.
   - The high standard deviation (11.99°C) suggests significant variability in temperature, which can impact gokart rentals.

2. **Humidity**:
   - The mean humidity is 58.25%, with values ranging from 0% to 98%.
   - This indicates that the dataset includes conditions from completely dry to nearly saturated air.
   - The wide range and high standard deviation (20.30%) show considerable variation in humidity levels.

3. **Wind Speed**:
   - The average wind speed is 1.73 m/s, with a range from calm (0 m/s) to strong winds (7.3 m/s).
   - This variability can affect the comfort and safety of gokart rentals.

4. **Visibility**:
   - The mean visibility is 1432.2 decameters, with a maximum of 2000 decameters.
   - Most visibility readings are at the maximum, indicating generally clear conditions.
   - Poor visibility conditions are rare but can affect gokart rentals.

5. **Dew Point Temperature**:
   - The mean dew point is 3.99°C, with values ranging from extremely dry (-29.8°C) to very humid (27.2°C).
   - This variation reflects different moisture levels in the air, which can impact comfort.

6. **Solar Radiation**:
   - The average solar radiation is low (0.57 MJ/m²), indicating that many observations are from times with little sunlight.
   - There is considerable variation (Std = 0.87 MJ/m²), which can affect the attractiveness of outdoor activities like gokart rentals.

7. **Precipitation**:
   - Both rainfall and snowfall have low averages, indicating infrequent occurrences.
   - However, extreme values (35 mm of rainfall and 8.8 cm of snowfall) show that there are some days with significant precipitation, likely affecting gokart rentals.

#### Gokart Rental Count:
- The average gokart rental count is 704.17, with a wide range from 0 to 3556 rentals per day.
- The high standard deviation (647.08) indicates significant variability in daily rentals, reflecting differing demand based on various factors such as weather, holidays, and weekends.

#### Seasonal and Temporal Patterns:
- The dataset covers observations throughout the year 2018, with data spread across all months.
- There are more observations in the middle of the year (mean month = 6.51), likely reflecting increased activity during warmer months.

### Insights and Potential Impacts on Gokart Rentals:
1. **Temperature**: 
   - Higher temperatures are likely associated with increased gokart rentals as outdoor activities become more appealing.
   - Extremely high or low temperatures might deter rentals due to discomfort.

2. **Humidity**: 
   - Moderate humidity levels might be ideal for rentals, whereas high humidity could make it uncomfortable for users.

3. **Wind Speed**: 
   - Calm to moderate wind speeds are likely more favorable for rentals. High winds could pose safety risks.

4. **Visibility**: 
   - Clear conditions (high visibility) are likely to encourage rentals as they are associated with pleasant weather.

5. **Dew Point Temperature**: 
   - Comfortable dew point temperatures (neither too dry nor too humid) are likely to positively impact rentals.

6. **Solar Radiation**: 
   - Sunny days with moderate solar radiation are likely to see higher rentals as people are more inclined to participate in outdoor activities.

7. **Precipitation**: 
   - Rain and snow negatively impact gokart rentals due to safety and comfort concerns.

8. **Temporal Factors**: 
   - Rentals might peak during weekends and holidays when people have more leisure time.
   - Seasonal trends, with more rentals in warmer months, reflecting the outdoor nature of the activity.

### Summary
The descriptive statistics and data analysis provide a comprehensive overview of the dataset, highlighting significant variability in weather conditions and gokart rental counts. These insights can help build a predictive model that accounts for the impact of various weather-related and temporal factors on gokart rentals. Understanding these patterns is crucial for optimizing operations and planning for varying demand.


## Handling Missing Values

Before proceeding with further analysis, it is crucial to handle any missing values in the dataset. Missing values can significantly impact the results of our analysis and the performance of machine learning models. We will check for missing values, visualize them, and handle them appropriately.

### Explanation

1. **Checking for Missing Values**:
   - We start by checking the dataset for any missing values using the `isnull().sum()` method. This will give us a count of missing values for each column.

2. **Visualizing Missing Values**:
   - To get a better understanding of the distribution of missing values, we use a heatmap. This visualization helps us quickly identify columns and rows with missing data.

3. **Handling Missing Values**:
   - There are several ways to handle missing values, including removing rows/columns with missing data or filling them with a specific value. In this case, we choose to fill the missing values with the mean of the respective column. This approach maintains the dataset's size and can be useful if the number of missing values is not excessively high.

4. **Verification**:
   - After handling the missing values, we verify that there are no more missing values left in the dataset by checking the sum of missing values again.

By handling missing values appropriately, we ensure that our subsequent analysis and modeling are not adversely affected by incomplete data.


In [5]:
# Check for missing values
data.isna().sum()


Date                        0
Hour                        0
Temperature(C)              0
Humidity(%)                 0
Wind speed (m/s)            0
Visibility (10m)            0
Dew point temperature(C)    0
Solar Radiation (MJ/m2)     0
Rainfall(mm)                0
Snowfall (cm)               0
Seasons                     0
Holiday                     0
Functioning Day             0
Gokart rental count         0
Day                         0
Month                       0
Year                        0
dtype: int64

### Conclusion on Missing Values

Based on the output from the code cell checking for missing values, we can conclude that the dataset does not contain any missing values for any of the features. This implies that the dataset is complete and has good data integrity, ensuring that all rows have entries for every feature, which is crucial for accurate analysis and modeling. Since there are no missing values, we can proceed with the analysis without needing to handle or impute missing data, simplifying the preprocessing steps. You can now continue with the next steps of your analysis with confidence in the completeness of your dataset.




# Cell 4: Data Visualization - Distributions
We visualize the distribution of key features (Temperature, Humidity, Wind Speed, Visibility, Dew Point Temperature, and Gokart Rental Count) using histograms. These plots help us understand the spread and central tendency of these variables.


In [ ]:
# Cell 4: Data Visualization - Distributions
plt.figure(figsize=(20, 15))

# Temperature Distribution
plt.subplot(2, 3, 1)
sns.histplot(data['Temperature(C)'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Temperature Distribution')

# Humidity Distribution
plt.subplot(2, 3, 2)
sns.histplot(data['Humidity(%)'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Humidity Distribution')

# Wind Speed Distribution
plt.subplot(2, 3, 3)
sns.histplot(data['Wind speed (m/s)'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Wind Speed Distribution')

# Visibility Distribution
plt.subplot(2, 3, 4)
sns.histplot(data['Visibility (10m)'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Visibility Distribution')

# Dew Point Temperature Distribution
plt.subplot(2, 3, 5)
sns.histplot(data['Dew point temperature(C)'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Dew Point Temperature Distribution')

# Gokart Rental Count Distribution
plt.subplot(2, 3, 6)
sns.histplot(data['Gokart rental count'].replace([np.inf, -np.inf], np.nan).dropna(), kde=True)
plt.title('Gokart Rental Count Distribution')

plt.tight_layout()
plt.show()


# Explanation of Distributions
- **Temperature Distribution**: Shows the range and common values for temperature, helping us see if the data is skewed or has outliers.
- **Humidity Distribution**: Displays the spread of humidity values, indicating typical humidity levels during the recorded periods.
- **Wind Speed Distribution**: Helps understand the common wind speeds and identify any unusual conditions.
- **Visibility Distribution**: Shows how visibility varies, which could affect gokart rentals.
- **Dew Point Temperature Distribution**: Provides insight into the dew point's variation, indicating overall moisture in the air.
- **Gokart Rental Count Distribution**: Crucial for understanding the target variable's spread, indicating how often different rental counts occur.

From these graphs, we can identify any potential outliers or skewness in the data that might need to be addressed.


### Detailed Explanation of Data Visualization - Distributions

#### Temperature Distribution


- **Description**: The temperature distribution shows the frequency of different temperature values in the dataset.
- **Insights**:
  - The distribution is approximately normal, with a peak around 10-15°C.
  - Temperatures range from around -17.5°C to 39.4°C.
  - There are some extreme values on both ends, indicating potential outliers.
  - The moderate temperatures (10-20°C) are the most common, suggesting that these conditions are prevalent in the dataset.
- **Impact on Gokart Rentals**:
  - Moderate temperatures are likely ideal for gokart rentals, as they are comfortable for outdoor activities.
  - Extreme temperatures, both high and low, might deter rentals due to discomfort.

#### Humidity Distribution


- **Description**: This plot shows the frequency of different humidity levels in the dataset.
- **Insights**:
  - The distribution is skewed towards higher humidity values, with a peak around 50-60%.
  - Humidity levels range from 0% to 98%.
  - High humidity values are quite common, indicating that the dataset includes many humid conditions.
- **Impact on Gokart Rentals**:
  - Moderate humidity levels are likely more comfortable for gokart rentals.
  - Extremely high humidity could make it uncomfortable for users, potentially reducing rentals.

#### Wind Speed Distribution


- **Description**: This plot shows the frequency of different wind speeds in the dataset.
- **Insights**:
  - The distribution shows a peak around 1-2 m/s, indicating light winds are common.
  - Wind speeds range from 0 to 7.3 m/s.
  - There are fewer observations with higher wind speeds, indicating that strong winds are less frequent.
- **Impact on Gokart Rentals**:
  - Calm to moderate wind speeds are likely more favorable for gokart rentals.
  - High winds could pose safety risks and reduce rentals.

#### Visibility Distribution


- **Description**: This plot shows the frequency of different visibility levels in the dataset.
- **Insights**:
  - Most visibility readings are at the maximum value of 2000 decameters, indicating generally clear conditions.
  - Visibility ranges from 27 to 2000 decameters.
  - Poor visibility conditions are rare but present, indicating occasional foggy or hazy conditions.
- **Impact on Gokart Rentals**:
  - Clear conditions (high visibility) are likely to encourage gokart rentals.
  - Poor visibility can deter rentals due to safety concerns.

#### Dew Point Temperature Distribution


- **Description**: This plot shows the frequency of different dew point temperatures in the dataset.
- **Insights**:
  - The distribution shows a peak around 0-10°C, indicating cool dew point temperatures are common.
  - Dew point temperatures range from -29.8°C to 27.2°C.
  - There are fewer observations at the extreme ends, indicating that extremely dry or humid air is less common.
- **Impact on Gokart Rentals**:
  - Comfortable dew point temperatures (neither too dry nor too humid) are likely to positively impact rentals.
  - Extreme dew point temperatures can affect user comfort.

#### Gokart Rental Count Distribution


- **Description**: This plot shows the frequency of different gokart rental counts in the dataset.
- **Insights**:
  - The distribution is right-skewed, with many days having fewer rentals and fewer days with very high rentals.
  - Rental counts range from 0 to 3556 per day.
  - The high variability in rental counts indicates differing demand based on various factors such as weather, holidays, and weekends.
- **Impact on Gokart Rentals**:
  - Understanding the distribution of rental counts helps in planning and managing gokart operations.
  - Identifying patterns in high and low rental days can help optimize staffing and resources.

### Summary
The distribution plots provide valuable insights into the dataset's characteristics and the potential impact of various weather conditions on gokart rentals. These insights can be used to build predictive models that account for the influence of weather and temporal factors on rental activity, helping to optimize operations and improve customer satisfaction.


# Cell 5: Data Visualization - Correlation Matrix
We convert categorical variables to numeric using one-hot encoding and then generate a correlation matrix. The heatmap visualization shows the correlation coefficients between all pairs of features, helping us identify which features are most strongly associated with gokart rentals.


In [ ]:
# Cell 5: Data Visualization - Correlation Matrix
# Convert categorical variables to numeric
data = pd.get_dummies(data, columns=['Seasons', 'Holiday', 'Functioning Day'], drop_first=True)

# Correlation Matrix
corr_matrix = data.corr()
plt.figure(figsize=(20, 15))
sns.heatmap(corr_matrix, annot=True, fmt='.2f', cmap='coolwarm', linewidths=0.5, linecolor='white')
plt.title('Detailed Feature Correlation Matrix')
plt.show()


# Explanation of Correlation Matrix
The correlation matrix helps us identify relationships between variables. Strong correlations (positive or negative) with the target variable (Gokart Rental Count) indicate which features are most influential. For instance, high temperature might correlate with higher rentals. This information guides feature selection and engineering.



### Detailed Explanation of Data Visualization - Correlation Matrix

#### Correlation Matrix


- **Description**: The correlation matrix shows the correlation coefficients between all pairs of features in the dataset. The heatmap visualizes these correlations with color intensity, where red indicates positive correlations and blue indicates negative correlations.
- **Insights**:
  - **Temperature and Gokart Rentals**: There is a moderate positive correlation between temperature and gokart rentals. This suggests that warmer weather is associated with higher gokart rentals.
  - **Humidity and Gokart Rentals**: There is a weak negative correlation between humidity and gokart rentals. Higher humidity might slightly deter rentals, but the effect is not strong.
  - **Wind Speed and Gokart Rentals**: There is a weak negative correlation between wind speed and gokart rentals. Higher wind speeds may slightly deter rentals.
  - **Visibility and Gokart Rentals**: There is a weak positive correlation between visibility and gokart rentals. Clearer conditions are associated with slightly higher rentals.
  - **Dew Point and Gokart Rentals**: There is a moderate positive correlation between dew point and gokart rentals. Higher dew point temperatures, indicating more humid conditions, are associated with higher rentals.
  - **Solar Radiation and Gokart Rentals**: There is a weak positive correlation between solar radiation and gokart rentals. Sunnier days might encourage more rentals.
  - **Rainfall and Gokart Rentals**: There is a weak negative correlation between rainfall and gokart rentals. Rainy conditions are likely to deter rentals.
  - **Snowfall and Gokart Rentals**: There is a weak negative correlation between snowfall and gokart rentals. Snowy conditions are likely to deter rentals.
  - **Hour and Gokart Rentals**: There is a weak positive correlation between the hour of the day and gokart rentals. This might indicate higher activity during certain hours.
  - **Day and Month**: There is no significant correlation with gokart rentals, suggesting that day and month alone do not strongly influence rental counts.

### Summary
The correlation matrix provides valuable insights into the relationships between different features and gokart rental counts. Understanding these correlations helps identify which weather conditions and temporal factors have the most significant impact on rentals. This information is crucial for building predictive models and optimizing operations based on weather forecasts and historical data.


# Exploratory Data Analysis

In this section, we will perform exploratory data analysis using pair plots and time series analysis. Pair plots will help us visualize the relationships between multiple features, while the time series analysis will help us understand the trend of gokart rentals over time. These techniques are essential for identifying patterns, correlations, and potential insights that can guide further analysis and model building.


In [ ]:
# Cell 6: Pair Plots
import seaborn as sns
import matplotlib.pyplot as plt

# Pair Plot for Key Features
plt.figure(figsize=(20, 15))
sns.pairplot(data[['Temperature(C)', 'Humidity(%)', 'Wind speed (m/s)', 'Visibility (10m)', 
                   'Dew point temperature(C)', 'Gokart rental count']])
plt.suptitle('Pair Plot of Key Features', y=1.02)
plt.show()


## Analysis of Pair Plots 

### Pair Plots

- **Description**: Pair plots display scatter plots for each pair of selected features along with the distribution of each feature on the diagonal. This helps in understanding the relationships and potential correlations between features.
- **Insights**:
  - **Temperature vs. Gokart Rentals**: 
    - **Observation**: There appears to be a positive correlation, with higher temperatures generally leading to more rentals.
    - **Learning**: Warmer weather is associated with increased outdoor activity, thus leading to higher gokart rentals. This suggests that temperature is an important factor in predicting rentals.
  - **Humidity vs. Gokart Rentals**: 
    - **Observation**: The scatter plot shows a slight negative correlation, indicating that higher humidity might deter rentals.
    - **Learning**: High humidity levels can make outdoor activities uncomfortable, thus reducing the number of rentals. This insight can be used to adjust rental predictions based on humidity levels.
  - **Wind Speed vs. Gokart Rentals**: 
    - **Observation**: The scatter plot suggests a weak negative correlation, with higher wind speeds slightly reducing rentals.
    - **Learning**: Windy conditions may make gokarting less enjoyable or safe, leading to fewer rentals. This highlights the need to consider wind speed in predictive models.
  - **Visibility vs. Gokart Rentals**: 
    - **Observation**: Higher visibility seems to be associated with more rentals, but the correlation is weak.
    - **Learning**: Clearer conditions likely encourage more rentals due to better driving conditions and overall visibility. However, since the correlation is weak, it may not be a primary factor.
  - **Dew Point vs. Gokart Rentals**: 
    - **Observation**: There is a positive correlation, with higher dew point temperatures (more humid conditions) being associated with more rentals.
    - **Learning**: This could indicate that moderate to high dew points, which reflect comfortable humidity levels, are favorable for rentals. Extreme dew points should still be considered cautiously.
- **Overall Learning**: The pair plots confirm the insights from the correlation matrix and provide a visual representation of the relationships between different features and gokart rentals. These plots help in understanding how various weather conditions impact gokart rental activity.



In [ ]:
# Cell 7: Time Series Analysis with Line Plots
import matplotlib.pyplot as plt
import seaborn as sns

# Line Plot for Gokart Rentals Over Time
plt.figure(figsize=(15, 8))
sns.lineplot(x=data['Date'], y=data['Gokart rental count'])
plt.title('Gokart Rentals Over Time')
plt.xlabel('Date')
plt.ylabel('Gokart Rentals')
plt.show()


## Time Series Analysis

### Time Series Analysis with Line Plots

The line plot visualizes the trend of gokart rentals over the entire observation period. It helps in identifying seasonal patterns, trends, and anomalies in the data.

- **Seasonal Trends**: 
  - **Observation**: There are clear seasonal trends, with rentals peaking during certain times of the year, likely corresponding to warmer months.
  - **Learning**: Gokart rentals are highly seasonal, with significant increases in rentals during warmer periods. This insight is crucial for seasonal planning and ensuring adequate resources during peak times.

- **Daily Fluctuations**: 
  - **Observation**: There are noticeable daily fluctuations in rentals, reflecting varying demand on different days.
  - **Learning**: The daily variations highlight the importance of considering day-to-day changes in rental activity. This can be influenced by factors such as weather, holidays, and weekends.

- **Holiday Effects**: 
  - **Observation**: Spikes in rentals can be observed around certain dates, potentially indicating holidays or special events.
  - **Learning**: Understanding the impact of holidays and events on rental activity can help in planning promotions and managing inventory. Special marketing strategies can be implemented around these times to maximize rentals.

- **Overall Learning**: The time series analysis reveals significant seasonal patterns in gokart rentals, with higher activity during warmer months. Understanding these trends can help in better planning and resource allocation, ensuring that supply meets demand during peak times.

### Summary
The pair plots and time series analysis provide a comprehensive view of the relationships between key features and gokart rentals. The pair plots confirm the correlations observed in the correlation matrix and offer visual insights into how weather conditions influence rentals. The time series analysis highlights the seasonal nature of gokart rentals, emphasizing the importance of considering temporal factors in any predictive modeling efforts. By understanding these patterns, we can improve operational planning, enhance customer satisfaction, and optimize resource allocation.


# Exploratory Data Analysis With Boxplot

In this section, we will perform exploratory data analysis using box plots. Box plots provide insights into the distribution of numerical features in the dataset and help identify outliers. This technique is essential for understanding the variability and central tendency of gokart rental counts in relation to various weather conditions.

To generate multiple box plots for all numerical features in the dataset, we will use the following numerical features: 'Temperature(C)', 'Humidity(%)', 'Wind speed (m/s)', 'Visibility (10m)', 'Dew point temperature(C)', 'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)', and 'Gokart rental count'.

The box plots visualize the distribution of various numerical features in the dataset. Each plot highlights the median, quartiles, and potential outliers for a specific feature. The median is represented by the middle line in each box, which corresponds to the 50th percentile of the feature. The interquartile range (IQR) is represented by the box, containing the middle 50% of the data (from the 25th percentile to the 75th percentile). The whiskers extend from the box to show the range of the data, excluding outliers, typically representing 1.5 times the IQR. Individual points beyond the whiskers indicate outliers, which are values that deviate significantly from the rest of the data.



In [ ]:
# Cell 8: Box Plots for Numerical Features
import matplotlib.pyplot as plt
import seaborn as sns

# List of numerical features
numerical_features = ['Temperature(C)', 'Humidity(%)', 'Wind speed (m/s)', 'Visibility (10m)',
                      'Dew point temperature(C)', 'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 
                      'Snowfall (cm)', 'Gokart rental count']

# Generate multiple box plots for all numerical features in the dataset
plt.figure(figsize=(15, 10))
for i, feature in enumerate(numerical_features):
    plt.subplot(3, 3, i+1)
    sns.boxplot(x=data[feature])
    plt.title(f'{feature} Boxplot')
plt.tight_layout()
plt.show()




## Box Plots of Numerical Features

**Temperature(C)**: The temperature distribution ranges from approximately -20°C to 40°C, with a median around 15°C. This wide range suggests the dataset covers multiple seasons with varying temperatures.

**Humidity(%)**: The humidity distribution ranges from 0% to nearly 100%, with a median around 60%. The wide range indicates varying humidity conditions, which can impact outdoor activities like gokart rentals.

**Wind speed (m/s)**: The wind speed has a median around 2 m/s, with several outliers beyond 4 m/s. Most wind speeds are low, but higher winds could affect the comfort and safety of gokarting.

**Visibility (10m)**: Visibility ranges mostly from 0 to 2000 decameters, with a median close to the maximum. High visibility is common, suggesting clear conditions most of the time.

**Dew point temperature(C)**: The dew point temperature ranges from -30°C to 30°C, with a median around 5°C. This indicates varying moisture levels in the air, affecting comfort.

**Solar Radiation (MJ/m2)**: Solar radiation ranges from 0 to about 3.5 MJ/m², with a median around 0.5 MJ/m². Low to moderate solar radiation is common, affecting the attractiveness of outdoor activities.

**Rainfall(mm)**: Rainfall values are mostly low, with several outliers indicating heavy rainfall days. Rainy days are less common, but they can significantly impact rentals.

**Snowfall (cm)**: Snowfall values are mostly zero, with some outliers indicating snow days. Snow is rare but can deter gokart rentals due to safety concerns.

**Gokart rental count**: The rental counts range widely, with a median around 500. There are several outliers with very high rental counts. High variability in rentals suggests that several factors, including weather and events, significantly impact rental activity.

### Summary
The box plots offer a detailed view of the distribution of numerical features in the dataset. They highlight the variability and central tendency of each feature and identify outliers. By understanding these distributions, we can better assess how different weather conditions and other factors influence gokart rentals, ultimately aiding in predictive modeling and operational planning.


# Comprehensive Conclusion

In this exploratory data analysis, we examined the gokart rental dataset to uncover patterns and relationships between various weather conditions and gokart rental counts. Our analysis employed several techniques, including visualizations, correlation matrices, box plots, and plots, to gain insights into the factors influencing gokart rentals.

The key findings from our analysis are as follows:

Temperature showed a positive correlation with gokart rentals. Higher temperatures generally led to an increase in rentals, likely due to the more favorable conditions for outdoor activities. The box plot of temperature ranges confirmed this trend, indicating that the median and variability of gokart rentals increase with warmer temperatures.

Humidity displayed a weak negative correlation with gokart rentals. High humidity levels can make outdoor activities uncomfortable, potentially deterring rentals.

Wind speed had a weak negative correlation with gokart rentals. Higher wind speeds might pose safety risks and reduce the enjoyment of gokarting, thereby reducing rentals.

Visibility was positively correlated with gokart rentals. Clear conditions are more conducive to outdoor activities, which likely leads to higher rental counts.

Dew point temperature showed a moderate positive correlation with gokart rentals. Comfortable humidity levels, indicated by moderate dew point temperatures, were associated with higher rentals.

There was a weak positive correlation between solar radiation and gokart rentals. Sunnier days might encourage more outdoor activities, leading to increased rentals.

Both rainfall and snowfall had negative correlations with gokart rentals. Adverse weather conditions like rain and snow deter outdoor activities due to safety and comfort concerns.

Time series analysis revealed significant seasonal patterns in gokart rentals. Rentals peaked during warmer months, emphasizing the importance of seasonal planning to ensure adequate resources during peak times. Daily fluctuations in rentals were also observed, reflecting varying demand influenced by factors such as weather, holidays, and weekends.

The exploratory data analysis provided valuable insights into how different weather conditions and temporal factors influence gokart rentals. Key findings indicate that temperature, visibility, and dew point temperature positively affect rental counts, while humidity, wind speed, rainfall, and snowfall have negative impacts. Seasonal and daily trends also play crucial roles in rental activity. Understanding these patterns can help optimize operational planning, enhance customer satisfaction, and improve resource allocation.

By integrating these insights into predictive models, gokart rental companies can better anticipate demand, plan for peak times, and develop targeted marketing strategies to maximize rentals during favorable conditions.
